File: mcp-code-intelligence-nodejs/src/orchestration/cache/cache-entry.ts
/**
 * CacheEntry — interface and helpers for adaptive token cache entries.
 */

export interface CacheEntry {
  tokens: Set<string>;
  toolName: string;
  score: number;
  registryHash: string;
  timestamp: string;
  hitCount: number;
  lastHit: string;
}

/** Create a new CacheEntry. */
export function createCacheEntry(
  tokens: Set<string>, toolName: string, score: number, registryHash: string
): CacheEntry {
  const now = nowIso();
  return { tokens, toolName, score, registryHash, timestamp: now, hitCount: 0, lastHit: now };
}

/** Record a cache hit on an entry. */
export function touchEntry(entry: CacheEntry): void {
  entry.hitCount++;
  entry.lastHit = nowIso();
}

/** Check if entry is stale (registry changed). */
export function isStale(entry: CacheEntry, currentHash: string): boolean {
  return entry.registryHash !== currentHash;
}

/** Serialize entry to JSON-compatible object. */
export function entryToJson(entry: CacheEntry): Record<string, any> {
  return {
    tokens: [...entry.tokens].sort(),
    tool_name: entry.toolName,
    score: entry.score,
    timestamp: entry.timestamp,
    hit_count: entry.hitCount,
    last_hit: entry.lastHit,
    tool_version: entry.registryHash,
  };
}

/** Deserialize entry from JSON object. */
export function entryFromJson(data: Record<string, any>): CacheEntry {
  return {
    tokens: new Set(data.tokens ?? []),
    toolName: data.tool_name ?? '',
    score: data.score ?? 0,
    registryHash: data.tool_version ?? '',
    timestamp: data.timestamp ?? nowIso(),
    hitCount: data.hit_count ?? 0,
    lastHit: data.last_hit ?? nowIso(),
  };
}

function nowIso(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');
}

File: backend/mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/orchestration/cache/CacheEntry.kt
/**
 * CacheEntry — data class for adaptive token cache entries.
 */
package com.codeintel.orchestration.cache

import kotlinx.serialization.Serializable
import java.time.Instant

@Serializable
data class CacheEntry(
    var tokens: Set<String>,
    val toolName: String,
    var score: Double,
    val registryHash: String,
    val timestamp: String = nowIso(),
    var hitCount: Int = 0,
    var lastHit: String = nowIso()
) {
    /** Record a cache hit. */
    fun touch() {
        hitCount++
        lastHit = nowIso()
    }

    /** Check if entry is stale (registry changed). */
    fun isStale(currentHash: String): Boolean = registryHash != currentHash

    /** Serialize to JSON-compatible map. */
    fun toMap(): Map<String, Any> = mapOf(
        "tokens" to tokens.sorted(),
        "tool_name" to toolName,
        "score" to score,
        "timestamp" to timestamp,
        "hit_count" to hitCount,
        "last_hit" to lastHit,
        "tool_version" to registryHash
    )

    companion object {
        /** Deserialize from JSON map. */
        fun fromMap(data: Map<String, Any?>): CacheEntry {
            val tokens = (data["tokens"] as? List<*>)?.mapNotNull { it?.toString() }?.toSet() ?: emptySet()
            return CacheEntry(
                tokens = tokens,
                toolName = data["tool_name"]?.toString() ?: "",
                score = (data["score"] as? Number)?.toDouble() ?: 0.0,
                registryHash = data["tool_version"]?.toString() ?: "",
                timestamp = data["timestamp"]?.toString() ?: nowIso(),
                hitCount = (data["hit_count"] as? Number)?.toInt() ?: 0,
                lastHit = data["last_hit"]?.toString() ?: nowIso()
            )
        }
    }
}

internal fun nowIso(): String = Instant.now().toString().replace(Regex("\\.\\d+Z$"), "Z")
